File: pkg/repoutils/current.go
package repoutils

import (
	"errors"
	"path/filepath"
	"strings"

	"github.com/fioncat/roxide/pkg/choice"
	"github.com/fioncat/roxide/pkg/context"
	"github.com/fioncat/roxide/pkg/db"
)

func MustGetCurrentRepo(ctx *context.Context) (*db.Repository, error) {
	repo, err := GetCurrentRepo(ctx)
	if err != nil {
		return nil, err
	}
	if repo == nil {
		return nil, errors.New("you are not in any repository")
	}

	return repo, nil
}

func GetCurrentRepo(ctx *context.Context) (*db.Repository, error) {
	dir := ctx.WorkDir
	for dir != "" {
		repo, err := getCurrentRepo(ctx, dir)
		if err != nil {
			return nil, err
		}
		if repo != nil {
			return repo, nil
		}
		dir = filepath.Dir(dir)
	}
	return nil, nil
}

func getCurrentRepo(ctx *context.Context, dir string) (*db.Repository, error) {
	id := ParseWorkspacePath(ctx, dir)
	if id == "" {
		repos, err := ctx.Database.QueryRepos(db.QueryRepositoryOptions{
			Path: &dir,
		})
		if err != nil {
			return nil, err
		}
		if len(repos) == 0 {
			return nil, nil
		}

		return repos[0], nil
	}

	repo, err := ctx.Database.GetRepo(id)
	if err != nil {
		if db.IsNotFound(err) {
			return nil, nil
		}
		return nil, err
	}

	return repo, nil
}

func ParseWorkspacePath(ctx *context.Context, dir string) string {
	if !strings.HasPrefix(dir, ctx.Config.Workspace) {
		return ""
	}

	path := strings.TrimPrefix(dir, ctx.Config.Workspace)
	path = strings.Trim(path, "/")

	fields := strings.Split(path, "/")
	if len(fields) < 3 {
		return ""
	}

	remote := fields[0]
	path = strings.Join(fields[1:], "/")

	owner, name := choice.ParseOwner(path)
	if owner == "" || name == "" {
		return ""
	}

	return db.BuildRepoID(remote, owner, name)
}

File: src/repo/current.rs
use std::sync::Arc;

use anyhow::{Result, bail};

use crate::config::context::ConfigContext;
use crate::db::repo::Repository;
use crate::debug;

pub fn get_current_repo(ctx: Arc<ConfigContext>) -> Result<Repository> {
    match get_current_repo_optional(ctx)? {
        Some(repo) => Ok(repo),
        None => bail!("you are not in any repository"),
    }
}

pub fn get_current_repo_optional(ctx: Arc<ConfigContext>) -> Result<Option<Repository>> {
    debug!(
        "[current] Get current repo, dir: {}",
        ctx.current_dir.display()
    );
    let db = ctx.get_db()?;

    if let Some((remote, owner, name)) =
        super::parse_workspace_path(&ctx.cfg.workspace, &ctx.current_dir)
    {
        debug!(
            "[current] Now in workspace repo, remote: {remote:?}, owner: {owner:?}, name: {name:?}"
        );
        let repo = db.with_transaction(|tx| tx.repo().get_optional(&remote, &owner, &name))?;
        debug!("[current] Current workspace repo: {repo:?}");
        return Ok(repo);
    }

    debug!("[current] Not in workspace repo, try by path");
    let path = format!("{}", ctx.current_dir.display());
    let repo = db.with_transaction(|tx| tx.repo().get_by_path_optional(&path))?;
    debug!("[current] Current path repo: {repo:?}");
    Ok(repo)
}

#[cfg(test)]
mod tests {
    use std::fs;
    use std::path::PathBuf;

    use crate::config::context;
    use crate::repo;

    use super::*;

    #[test]
    fn test_current_repo() {
        let ctx = context::tests::build_test_context(
            "current_repo",
            Some(PathBuf::from("/path/to/nvimdots")),
        );
        let repo = get_current_repo(ctx).unwrap();
        assert_eq!(
            repo,
            Repository {
                remote: "github".to_string(),
                owner: "fioncat".to_string(),
                name: "nvimdots".to_string(),
                path: Some("/path/to/nvimdots".to_string()),
                sync: true,
                pin: true,
                last_visited_at: 3333,
                visited_count: 0,
                new_created: false,
            }
        );
    }

    #[test]
    fn test_current_repo_workspace() {
        let workspace = "./tests/current_repo_workspace";
        repo::ensure_dir(workspace).unwrap();
        let workspace = fs::canonicalize(workspace).unwrap();
        let repo_path = workspace
            .join("workspace")
            .join("github")
            .join("fioncat")
            .join("roxide")
            .join("src");
        let ctx = context::tests::build_test_context("current_repo_workspace", Some(repo_path));
        let repo = get_current_repo(ctx).unwrap();
        assert_eq!(
            repo,
            Repository {
                remote: "github".to_string(),
                owner: "fioncat".to_string(),
                name: "roxide".to_string(),
                path: None,
                sync: true,
                pin: true,
                last_visited_at: 2234,
                visited_count: 20,
                new_created: false,
            }
        )
    }
}
